osh/cmd_parse.py

```diff
@@ -482,12 +483,18 @@ def ParseRedirect(self):
     assert self.c_kind == Kind.Redir, self.cur_word
     op_tok = cast(Token, self.cur_word)  # for MyPy
 
-    # For now only supporting single digit descriptor
-    first_char = op_tok.val[0]
-    if first_char.isdigit():
-      fd = int(first_char)
+    fdspec = ''
+    if op_tok.val[0] == '{':
+      index = op_tok.val.find('}')
+      if index < 0:
+        p_die('Invalid token after redirect operator', word=self.cur_word)
+      fdspec = op_tok.val[:index+1]
     else:
-      fd = runtime.NO_SPID
+      index = 0
+      while index < len(op_tok.val) and op_tok.val[index].isdigit():
```

Comment:
Important issue: I think we need a limit on user descriptors so they don't collide with descriptors the shell uses?  Mentioned here:

https://www.aosabook.org/en/bash.html

> The other complication is one bash brought on itself. Historical versions of the Bourne shell allowed the user to manipulate only file descriptors 0-9, reserving descriptors 10 and above for the shell's internal use. Bash relaxed this restriction, allowing a user to manipulate any descriptor up to the process's open file limit. This means that bash has to keep track of its own internal file descriptors, including those opened by external libraries and not directly by the shell, and be prepared to move them around on demand. This requires a lot of bookkeeping, some heuristics involving the close-on-exec flag, and yet another list of redirections to be maintained for the duration of a command and then either processed or discarded.

Are you using descriptors over 9?  I haven't seen shell scripts using more, and POSIX only guarantees up to 9 I think.

Maybe we can limit it to  99 instead of 9 ?  But if there's no usage I would keep it at 9 until we encounter it.